File: 994_Rotting_Oranges/994_1.py
from typing import List

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rotten_list = []
        fresh_num = 0
        row_len = len(grid)
        col_len = len(grid[0])
        for i in range(row_len):
            for j in range(col_len):
                if grid[i][j] == 2:
                    rotten_list.append([i, j])

        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        rtime = 0
        while rotten_list:
            new_rot_list = []
            for x0, y0 in rotten_list:
                for k in range(4):
                    x = x0 + dx[k]
                    y = y0 + dy[k]
                    if 0 <= x < row_len and 0 <= y < col_len and grid[x][y] == 1:
                        grid[x][y] = 2
                        new_rot_list.append([x, y])
                #        fresh_num -= 1
            
            if not new_rot_list: break
            rotten_list =  new_rot_list[:]
            rtime += 1

        for row in grid:
            for i in row:
                if i == 1:
                    return -1
        return rtime


            
if __name__ =="__main__":
    aa = [[2,1,1],[1,1,0],[0,1,1]]
    s = Solution()
    print(s.orangesRotting(aa))

File: 994_Rotting_Oranges/994_1.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Queue;
import java.util.ArrayDeque;


class Solution {
	public int orangesRotting(int[][] grid) {
		int[] dx = new int[]{-1, 0, 1, 0};
		int[] dy = new int[]{0, -1, 0, 1};
		int row = grid.length;
		int col = grid[0].length;
		Queue<int []> queue = new ArrayDeque<>();
		for (int i = 0; i < row; ++i) {
			for (int j = 0; j < col; ++j)
				if (grid[i][j] == 2)
					queue.add(new int[]{i, j, 0});
		}

		int[] head = new int[3];
		while (!queue.isEmpty()) {
			head = queue.poll();
			for (int i = 0; i < 4; ++i) {
				int x = head[0] + dx[i];
				int y = head[1] + dy[i];
				if (x >= 0 && x < row && y >= 0 && y < col && grid[x][y] == 1) {
					grid[x][y] = 2;
					queue.add(new int[]{x, y, head[2] + 1});
				}
			}
		}
		for (int i = 0; i < row; ++i) {
			for (int j = 0; j < col; ++j)
				if (grid[i][j] == 1)
					return -1;
		}
		return head[2];
	}
}

public class leetcode {
	public static void main(String[] args) {
		int[][] aa = {{2,1,1},{1,1,0},{0,1,1}};
    	Solution s = new Solution();
    	System.out.println(s.orangesRotting(aa));
	}
}

